### PR TITLE
Generate release notes

### PR DESCRIPTION
**Pull request by petersutter**
Generate release notes from specially tagged quotes in pull request descriptions.

Introducing new method `generate_release_notes` to generate release notes (inspired by https://github.com/kubernetes/release/blob/master/relnotes).

Flow is:

1. calculate range since last release
2. get commits in range and gather pull request numbers from merge commits 
3. scrape description of pull requests with code block

> \`\`\` \<category\> \<target_group\>
> \<release note text\>
> \`\`\`

**category**: improvement|noteworthy
**target_group**: user|operator

4. build release note text as markdown string

### CODE REVIEW

File: cli/githubutil.py [hunk 1]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+):
+    repo = git.Repo(existing_dir(repo_dir))
+    if repo.active_branch.name != repository_branch:
+        fail('need to switch to branch {branch}. Current active branch: {active_branch}'.format(branch=repository_branch, active_branch=repo.active_branch.name))
+
+    if not commit_range:
+        commit_range = calculate_range(repo, helper)
+    pr_numbers = fetch_pr_numbers_in_range(repo, commit_range)
+    release_note_objs = fetch_release_notes_from_prs(helper, pr_numbers)
+    release_notes = build_markdown(release_note_objs)
```
Comment (ccwienk) [inline cli/githubutil.py]: rename to release_notes_str or sth similar to express the difference between the objects and the resulting release-note-"text"

File: cli/githubutil.py [hunk 2]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+        return release_note_lines
+
+    def release_notes_for_category(
+        category: Category,
+        release_note_objs: list
+    ) -> list:
+        rn_lines = list()
+        rn_lines_category = list()
+        for group in groups:
+            rn_lines_category.extend(release_notes_for_group(category=category, group=group, release_note_objs=release_note_objs))
```
Comment (ccwienk) [inline cli/githubutil.py]: line length..

File: cli/githubutil.py [hunk 3]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+    ) -> list:
+        release_note_objs = _.filter(release_note_objs, lambda release_note: category.identifier == release_note.category_id and group.identifier == release_note.group_id)
+
+        release_note_lines = list()
+        if release_note_objs:
+            release_note_lines.append(group.title)
+            for release_note in release_note_objs:
+                for i, rls_note_line in enumerate(release_note.text.splitlines()):
+                    if i == 0:
+                        release_note_lines.append('* {rls_note_line} (#{pr_num}, @{user})'.format(pr_num=release_note.pr_number, user=release_note.user_login, rls_note_line=rls_note_line))
```
Comment (ccwienk) [inline cli/githubutil.py]: length

File: cli/githubutil.py [hunk 4]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+        return rn_lines
+
+    release_note_lines = list()
+    for category in categories:
+        release_note_lines.extend(release_notes_for_category(category, release_note_objs=release_note_objs))
+
+    if release_note_lines:
+        return '\n'.join(release_note_lines)
+    else: # fallback
+        return 'release notes'
```
Comment (ccwienk) [inline cli/githubutil.py]: should this not rather be a value like 'no release notes available'?

File: cli/githubutil.py [hunk 5]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+    return release_notes
+
+def extract_release_notes(
+    pr_number: int,
+    text: str,
+    user_login: str
+) -> list:
+    release_notes = list()
+
+    quotes = re.findall(r"``` *(improvement|noteworthy)( (user|operator)?)?.*?\n(.*?)\n```", text, re.MULTILINE | re.DOTALL)
```
Comment (ccwienk) [inline cli/githubutil.py]: line-length..

File: cli/githubutil.py [hunk 6]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+
+        pr_number = pr_dict['number']
+        if not str(pr_number) in pr_numbers_in_range:
+            continue
+
+        release_notes_pr = extract_release_notes(pr_number=pr_number, text=pr_dict['body'], user_login=_.get(pr_dict, 'user.login'))
+        if not release_notes_pr:
+            continue
+
+        release_notes.extend(release_notes_pr)
```
Comment (ccwienk) [inline cli/githubutil.py]: why not yield?

File: cli/githubutil.py [hunk 7]
```diff
@@ -71,6 +85,213 @@ def assign_github_team_to_repo(
...
+
+        text = quote[3]
+        if not text or 'none' == text.lower():
+            continue
+
+        category = quote[0]
+        group = quote[2] or 'user'
+
+        release_notes.append(ReleaseNote(category_id=category, group_id=group, text=text, pr_number=pr_number, user_login=user_login))
+    return release_notes
```
Comment (ccwienk) [inline cli/githubutil.py]: yield

File: cli/githubutil.py [hunk 8]
```diff
@@ -79,28 +300,15 @@ def release_and_prepare_next_dev_cycle(
     repository_branch: str,
     repository_version_file_path: str,
     release_version: str,
-    release_notes: str,
+    release_notes: str=None,
```
Comment (ccwienk) [inline cli/githubutil.py]: we do not have to make this backward-compatible

File: github/util.py [hunk 9]
```diff
@@ -258,6 +262,10 @@ def create_tag(
             tagger=author
         )
 
+    def get_latest_release(self):
```
Comment (ccwienk) [inline github/util.py]: rm dead code

File: github/util.py [hunk 10]
```diff
@@ -308,6 +316,14 @@ def release_versions(self):
                 yield semver.parse_version_info(release.tag_name)
             except ValueError: pass # ignore
 
+    def release_tags(self):
+        return _.map(self.repository.releases(), 'tag_name')
```
Comment (ccwienk) [inline github/util.py]: vanilla python3 alternatives:
return (release.tag_name for release in self.repository.releases())
return map(lambda release: release.tag_name, self.repository.releases())

File: cli/githubutil.py (hunk 6 above)
Comment (petersutter) [reply cli/githubutil.py]: because later I have to filter on the list

File: cli/githubutil.py (hunk 7 above)
Comment (petersutter) [reply cli/githubutil.py]: see comment from above (I have to filter on the list)

**Comment by petersutter** [comment]
FYI there is still one false positive linting error:

```
************* Module github.release_notes
E:150,67: Undefined variable 'item' (undefined-variable)
```
https://github.com/PyCQA/pylint/issues/760

**Comment by ccwienk** [comment]
@petersutter : as for the false positive linting error, please disable said warning (grep for 'pylint disable' for examples in our codebase)

File: cli/githubutil.py [hunk 11]
```diff
@@ -124,7 +347,7 @@ def release_and_prepare_next_dev_cycle(
     )
     helper.create_tag(
         tag_name=release_version,
-        tag_message=release_notes,
+        tag_message="Release " + release_version, # TODO
```
[the remark this answers is not in the document]
Comment (petersutter) [reply cli/githubutil.py]: https://github.com/gardener/cc-utils/pull/70

**Comment by AndreasBurger** [comment]
There seem to be issues with git-hub here, I'll close this one and reopen it.

**Pull request closed without merging**